### DISCUSSION TITLE
byte order and floats, what am I doing wrong?

### DISCUSSION THREAD

**Discussion by haisamido** [CCSDSPy/ccsdspy#110] [Q&A] [answered]
**_byte order and floats, what am I doing wrong?_**

I have my own python code, below, that successfully returns the `float` for `MNEMONIC`, whose binary byte order  is `3412`.

```python
mnemonic_ba = bytearray([ byte_arr[offset+36+1], byte_arr[offset+36+0], byte_arr[offset+36+3], byte_arr[offset+36+2] ])
MNEMONIC    = (struct.unpack('<f', mnemonic_ba))[0]
```

However, when I use ccsdspy, with the below definition for `MNEMONIC` with `byte_order` little or big
```python
PacketField(name='MNEMONIC',    data_type='float', bit_length=32, byte_order='little'),
```
the values that are returned are wrong and are either extremely large or extremely small float values. However, all `uint`'s that I am returning by ccsdspy are correctly returned. _The issue I'm having is only with floats._

**Comment by ddasilva** [comment]
Maybe I'm missing something, but I haven't encountered 3412 before. The little/big endian orders are 4321 and 1234... is there a name for this 3412? Does it happen a lot in your use cases?

What you could do in CCSDSPy is parse two uint fields, one for bytes 34 and one for bytes 12, and then write a [custom converter](https://docs.ccsdspy.org/en/latest/user-guide/converters.html) to combine them. Something like this:

```python

from ccsdspy import FixedLength, converters

class CustomConverter(converters.Converter):
    def __init__(self):
        pass
    def convert(self, first, second):
        # combine them here

pkt = FixedLength([
    PacketField(name="first", data_type="uint", bit_length=16),
    PacketField(name="second", data_type="uint", bit_length=16),
])
pkt.add_converted_field(
    ("first", "second"),
    "MyField_Converted",
    CustomConverter()
)

result = pkt.load("MyCCSDS.bin")

print(result["MyField_Converted"])
```

I haven't tested this code to implement `# combine them here`, but I think you could do something like this:

```python
    def convert(self, first, second):
        combined = np.array([first, second], dtype=np.uint16).T.flatten()
        combined.dtype = np.float32
        return combined
```

**Comment by haisamido** [comment]
3412 is very common where I work. It's the dominant byte order. I think
permitting the option of providing numeric representation of byte order
would be very useful.

**Reply by ddasilva** [reply]
Understood. Do you have any public data in this byte order I can write a test case around?

It shouldn't be too hard to write a patch to accept byte_order="3412" for arbitrary strings like that.

**Reply by haisamido** [reply]
@ddasilva  thank you for your reply. _**Check your email when you have a chance**_. Here are all the byte orders that have to be decoded. Some are clearly little endian, while others are big endian, but others are something else, like `2143`, `3412` and `78563412`. It would be neat to be able to provide the below names to cover all cases. If they are implemented I can test them at my end.  

BTW, there is also `bit order` within the byte, which I've never seen in binary data, but is part of the [XML Telemetric and Command Exchange (XTCE)](https://www.omg.org/spec/XTCE/1.2/PDF) specification.

```bash
1         
12        
123       
1234      
123456    
1234567   
12345678  
21        
2143      
321       
3412      
4321      
78563412  
87654321
```

**Reply by ddasilva** [reply]
Hi @haisamido,

I implemented this for you in this branch `complex-byte-order`. It works when the bit length is divisible by 8, and you can just pass `byte_order="3412"` when defining a field. Can you run `pip install .` on that branch and let me know if it works? 

I designed it around test cases I wrote with synthetic test data. The synthetic data uses uints and all the XTCE orderings you listed above. If the code works for you, I'll add the proper documentation and push it to the main branch.

https://github.com/CCSDSPy/ccsdspy/tree/complex-byte-order

Best,
Daniel

**Reply by haisamido** [reply]
Thank you @ddasilva I'll let you know how testing goes.

**Reply by haisamido** [reply]
@ddasilva I successfully installed from the `complex-byte-order` branch ; however, when I use `byte_order='3412'` I get the following error:

```bash
File "/Users/hido/development/github.com/haisamido/ccsds/ccsds.py", line 201, in <module>
    result = pkt.load(fileName, include_primary_header=True)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/ccsdspy/packet_types.py", line 169, in load
    packet_arrays = _load(
                    ^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/ccsdspy/packet_types.py", line 640, in _load
    field_arrays = _apply_post_byte_reoderings(field_arrays, orig_fields)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/ccsdspy/packet_types.py", line 692, in _apply_post_byte_reoderings
    field_arrays[field._name][i] = _do_array_byte_reordering(
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/ccsdspy/packet_types.py", line 708, in _do_array_byte_reordering
    array_bytes.dtype = np.uint8
    ^^^^^^^^^^^^^^^^^
AttributeError: attribute 'dtype' of 'numpy.generic' objects is not writable
```

**Reply by ddasilva** [reply]
Thanks, I'll look into it. I wrote my tests around one type of field and it might require some additional handling to work with other types of arrays/fields. Should be quick to fix.

**Reply by ddasilva** [reply]
Can you send your packet definition or at least the field definition you're using?

**Reply by haisamido** [reply]
Sorry I forgot to include that initially:

` PacketField(name='FOOBAR', data_type='float', bit_length=32, byte_order='3412'),`

**Reply by ddasilva** [reply]
Thank you. I was able to reproduce this exception, added tests for it, and pushed a patch. Please run your code again and let me know if we're doing better.

**Reply by haisamido** [reply]
Thank you. The exception is gone with the new patch; however, the results still don't look correct. The float values that are being returned for the below fields

```python
PacketField(name='FOOBAR1', data_type='float', bit_length=32, byte_order='3412'),
PacketField(name='FOOBAR2', data_type='float', bit_length=32, byte_order='3412'),
PacketField(name='FOOBAR3', data_type='float', bit_length=32, byte_order='3412'),
PacketField(name='FOOBAR4', data_type='float', bit_length=32, byte_order='3412'),
```
are summarized below

```tsv
#name, expected float, returned float
FOOBAR1,-59.253719, -2205008400.0
FOOBAR2,65.153366, -3.8839992e-35
FOOBAR3,-37.579803,-3.955661e-19
FOOBAR4,64.456299,-4.9850678e-05
```
BTW the below returns the expected `uint` value

```python
PacketField(name='FOOBARCNT', data_type='uint',  bit_length=16, byte_order='21')
``` 
of `64960` **_but_** `byte_order='21'` should really be `byte_order='12'` to return the correct value. So it seems things are reversed? While `byte_order='big'` returns the correct and expected value of `64960`

**Reply by haisamido** [reply]
@ddasilva  when I used `byte_order='2134'` the float values are much closer to correct ones, but not really within precision. However, `byte_order='3412'` should be the one that's returning the correct values.

```tsv
#name, expected float, returned float
FOOBAR1,-59.253719, -59.45216
FOOBAR2,65.153366, 65.262314
FOOBAR3,-37.579803,-37.679996
FOOBAR4,64.456299,64.31428
```

**Reply by haisamido** [reply]
With `byte_order='2143'` the returned results are correct, and with the desired precision, while `byte_order='3412'` should be the one that's returning the correct results. We're getting closer :-) That's the power of the integer representation of byte_order.

**Reply by ddasilva** [reply, marked as answer]
Great! I changed two lines (basically added `byte_order.reverse()` to the library and test data)... let me know if it is better now.

**Reply by haisamido** [reply]
worked like a charm! Thank you @ddasilva :-) When do you think this branch will be part of main?

**Reply by ddasilva** [reply]
Great! are you okay working off that branch for now? I will find some time to write the proper documentation and add some additional tests later this week. 

Can you also try with your 6 byte floats and let me know what happens? 6 byte floats are a little more dicey because it's not a natural length for floats (it's not clear to me what you expect to happen).

**Reply by haisamido** [reply]
I don't have actual data for 6 byte floats.

**Reply by ddasilva** [reply]
Sounds good, thanks!

**Reply by ddasilva** [reply]
Merged into main, will be in the next release.

**Reply by haisamido** [reply]
Fantastic @ddasilva. Thank you for your amazing support.
